### PR TITLE
feat: History API (client)

### PR DESCRIPTION
**Pull request by cachho**
This PR adds chat history. It's to be seen as an API to add history in any way you want.

The history could be saved and retrieved from a client, a database or anything you want.

It's 100% optional and can be used without.

Done:
- integration into the template
- use with query, chat and dry run.

Resolves: #113, #198

### CODE REVIEW

**Comment by cachho** [comment]
Test:

*In this test I ask 'Can you explain that in more detail?', a question that cannot be answered without conversation history.* 

```python

app = App()

app.add_local('text', "lorem ipsum")

print("##### Without history #####")
response = app.query("Can you explain that in more detail?")
print(response)

history = ['User: Are dolphins fish?', 'Bot: Dolphins are not actually fish.']
config = QueryConfig(history=history)
print("##### With history #####")
print(app.dry_run("Can you explain that in more detail?", config))
response = app.query("Can you explain that in more detail?", config)
print("Answer: ", response)

```

result:

```
Successfully saved lorem ipsum. Total chunks count: 1
##### Without history #####
I'm sorry, but without any specific context or information, I am unable to provide a more detailed explanation.
##### With history #####

  Use the following pieces of context to answer the query at the end.
  If you don't know the answer, just say that you don't know, don't try to make up an answer.
  I will provide you with our conversation history. 

  lorem ipsum

  History: ['User: Are dolphins fish?', 'Bot: Dolphins are not actually fish.']

  Query: Can you explain that in more detail?

  Helpful Answer:

Answer:  Dolphins are not fish, but rather mammals. They belong to a group of marine mammals called cetaceans, which also includes whales and porpoises. Unlike fish, dolphins are warm-blooded, have lungs, and give live birth to their young. They also nurse their offspring with milk, just like other mammals. Dolphins have adapted to life in the water with streamlined bodies, a dorsal fin, and a tail fluke for swimming. They breathe through a blowhole on top of their heads, which allows them to come up to the surface to take in air. Dolphins are highly intelligent creatures known for their social behavior, communication skills, and acrobatic abilities.
```

Test successful.

**Comment by cachho** [comment]
refactored chat to use the new API. this allowed the deletion of `generate_chat_prompt`, because it can now use the regular `generate_prompt`. This improves maintainability.

Test:

```python
response = app.chat("My name is John Doe. Can you remember that?")
print("Answer: ", response)
response = app.chat("What is my name?")
print("Answer: ", response)
```

Result:
````
Answer:  Yes, I will remember that your name is John Doe.
(History: Human: My name is John Doe. Can you remember that?
AI: Yes, I will remember that your name is John Doe.)
Answer:  Your name is John Doe.
```

File: embedchain/config/QueryConfig.py [hunk 1]
```diff
@@ -13,25 +13,73 @@
...
+  Query: $query
+
+  Helpful Answer:
+"""
+
 DEFAULT_PROMPT_TEMPLATE = Template(DEFAULT_PROMPT)
+DEFAULT_PROMPT_WITH_HISTORY_TEMPLATE = Template(DEFAULT_PROMPT_WITH_HISTORY)
 query_re = re.compile(r"\$\{*query\}*")
 context_re = re.compile(r"\$\{*context\}*")
+history_re = re.compile(r"\$\{*context\}*")
```
Comment (ianupamsingh) [inline embedchain/config/QueryConfig.py:34]: Typo: this should be `history` instead of `context`

Comment (cachho) [reply embedchain/config/QueryConfig.py:34]: good catch

**Review by taranjeet — APPROVED** [review]

**Pull request merged**